<a href="https://colab.research.google.com/github/MarioAlmela/03MIAR__Algoritmos-de-Optimizacion/blob/main/Trabajo_Practico_Mario_Almela_Rubio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de Optimización - Trabajo Práctico
- **Nombre y Apellidos:** Mario Almela Rubio
- **Github Url:** https://github.com/MarioAlmela/03MIAR__Algoritmos-de-Optimizacion
- **Problema elegido:**  1. Sesiones de doblaje
- **Descripción del problema:** Se precisa coordinar el doblaje de una película. Los actores del doblaje deben coincidir en las
tomas en las que sus personajes aparecen juntos en las diferentes tomas. Los actores de
doblaje cobran todos la misma cantidad por cada día que deben desplazarse hasta el estudio de
grabación independientemente del número de tomas que se graben. No es posible grabar más
de 6 tomas por día. El objetivo es planificar las sesiones por día de manera que el gasto por los
servicios de los actores de doblaje sea el menor posible.

In [ ]:
# Importación de librerías
import numpy as np
import pandas as pd
import itertools
import math
import random
import time

## Datos
Los datos iniciales para este problema son:

In [ ]:
initial_config = np.array([[1,1,1,1,1,0,0,0,0,0],
                           [0,0,1,1,1,0,0,0,0,0],
                           [0,1,0,0,1,0,1,0,0,0],
                           [1,1,0,0,0,0,1,1,0,0],
                           [0,1,0,1,0,0,0,1,0,0],
                           [1,1,0,1,1,0,0,0,0,0],
                           [1,1,0,1,1,0,0,0,0,0],
                           [1,1,0,0,0,1,0,0,0,0],
                           [1,1,0,1,0,0,0,0,0,0],
                           [1,1,0,0,0,1,0,0,1,0],
                           [1,1,1,0,1,0,0,1,0,0],
                           [1,1,1,1,0,1,0,0,0,0],
                           [1,0,0,1,1,0,0,0,0,0],
                           [1,0,1,0,0,1,0,0,0,0],
                           [1,1,0,0,0,0,1,0,0,0],
                           [0,0,0,1,0,0,0,0,0,1],
                           [1,0,1,0,0,0,0,0,0,0],
                           [0,0,1,0,0,1,0,0,0,0],
                           [1,0,1,0,0,0,0,0,0,0],
                           [1,0,1,1,1,0,0,0,0,0],
                           [0,0,0,0,0,1,0,1,0,0],
                           [1,1,1,1,0,0,0,0,0,0],
                           [1,0,1,0,0,0,0,0,0,0],
                           [0,0,1,0,0,1,0,0,0,0],
                           [1,1,0,1,0,0,0,0,0,1],
                           [1,0,1,0,1,0,0,0,1,0],
                           [0,0,0,1,1,0,0,0,0,0],
                           [1,0,0,1,0,0,0,0,0,0],
                           [1,0,0,0,1,1,0,0,0,0],
                           [1,0,0,1,0,0,0,0,0,0]])

# Creamos y configuramos el dataframe que contendrá los datos
data = pd.DataFrame(initial_config, index=np.arange(30), columns=['Actor ' + str(x) for x in np.arange(1, 11)])
data.index.name = 'Toma'
data['Total'] = np.sum(data, axis=1)
display(data)

Actor 1  Actor 2  Actor 3  Actor 4  Actor 5  Actor 6  Actor 7  Actor 8  \
Toma                                                                           
0           1        1        1        1        1        0        0        0   
1           0        0        1        1        1        0        0        0   
2           0        1        0        0        1        0        1        0   
3           1        1        0        0        0        0        1        1   
4           0        1        0        1        0        0        0        1   
5           1        1        0        1        1        0        0        0   
6           1        1        0        1        1        0        0        0   
7           1        1        0        0        0        1        0        0   
8           1        1        0        1        0        0        0        0   
9           1        1        0        0        0        1        0        0   
10          1        1        1        0        1        0        0        1   
11          1        1        1        1        0        1        0        0   
12          1        0        0        1        1        0        0        0   
13          1        0        1        0        0        1        0        0   
14          1        1        0        0        0        0        1        0   
15          0        0        0        1        0        0        0        0   
16          1        0        1        0        0        0        0        0   
17          0        0        1        0        0        1        0        0   
18          1        0        1        0        0        0        0        0   
19          1        0        1        1        1        0        0        0   
20          0        0        0        0        0        1        0        1   
21          1        1        1        1        0        0        0        0   
22          1        0        1        0        0        0        0        0   
23          0        0        1        0        0        1        0        0   
24          1        1        0        1        0        0        0        0   
25          1        0        1        0        1        0        0        0   
26          0        0        0        1        1        0        0        0   
27          1        0        0        1        0        0        0        0   
28          1        0        0        0        1        1        0        0   
29          1        0        0        1        0        0        0        0   

      Actor 9  Actor 10  Total  
Toma                            
0           0         0      5  
1           0         0      3  
2           0         0      3  
3           0         0      4  
4           0         0      3  
5           0         0      4  
6           0         0      4  
7           0         0      3  
8           0         0      3  
9           1         0      4  
10          0         0      5  
11          0         0      5  
12          0         0      3  
13          0         0      3  
14          0         0      3  
15          0         1      2  
16          0         0      2  
17          0         0      2  
18          0         0      2  
19          0         0      4  
20          0         0      2  
21          0         0      4  
22          0         0      2  
23          0         0      2  
24          0         1      4  
25          1         0      4  
26          0         0      2  
27          0         0      2  
28          0         0      3  
29          0         0      2

## ¿Cuántas posibilidades hay sin tener en cuenta las restricciones? ¿Cuántas posibilidades hay teniendo en cuenta todas las restricciones?

Para responder a esta pregunta, primero se deben identificar las restricciones que plantea el problema. En este caso, serían:
- **Los actores deben coincidir para grabar las toma en las que sus personajes aparecen juntos:** Esta es la restricción más obvia, y ya está expresada en la forma de los datos, que nos indican qué actor participa en qué tomas.
- **No es posible grabar más de 6 tomas por día:** La segunda restricción más obvia es que el número de tomas que se puede grabar por día es limitado. De no existir esta restricción se podrían grabar todas las tomas en un solo día.
- **Cada día deberían grabarse cuantas más tomas mejor:** Esta restricción no se especifica en el enunciado, pero de no tenerse en cuenta, las soluciones obtenidas no se acercarían nunca al óptimo. Dado que cada actor cobra por día de grabación, grabar una única toma por día (o no completar la cuota de tomas por día) es por definición subóptimo.

Teniendo estas restricciones en cuenta, si no existiera ninguna de ellas, las posibilidades serían infinitas ($∞$), ya que se podría grabar la participación de cualquier actor en cualquier día existente. Dado que no tenemos necesidad de grabar un mínimo de tomas al día, habría días en los que se grabarían 0 tomas.

Teniendo en cuenta la primera restricción únicamente, más la lógica de grabar al menos una toma al día, los posibles días en los que se puede completar la grabación van de 1 único día (grabando todas las tomas en una sola sesión) a 30 días (grabando solo una toma al día). En ese sentido, la cantidad de combinaciones posibles asciende a $30^{30}$.

Por último, teniendo en cuenta todas las restricciones, el problema consiste en asignar a cada toma un día concreto, tratando de llenar siempre la cuota máxima de 6 tomas por día. Las tomas que caigan en el mismo día pueden grabarse en cualquier orden, debido a que los actores van a cobrar lo mismo siempre que las tomas se graben en el mismo día. Para este problema en concreto, esto nos da el siguiente conjunto:

$A = \{1,1,1,1,1,1,2,2,2,2,2,2,3,3,3,3,3,3,4,4,4,4,4,4,5,5,5,5,5,5\}$

Las soluciones serían las permutaciones con repetición sobre el conjunto $A$, lo que nos da un número de posibilidades:

$P = \displaystyle \frac{30!}{6! 6! 6! 6! 6!} = \displaystyle \frac{30!}{720^5} = 1370874167589326400$

## Modelo para el espacio de soluciones ¿Cuál es la estructura de datos que mejor se adapta al problema?

En primera instancia, debido a lo explicado en el apartado anterior, el modelo de soluciones inicial consiste en una lista de números enteros. Esta lista tiene el mismo tamaño que la cantidad de tomas disponibles para grabar, y representa el día asignado a cada toma. El valor de la posición $n$ corresponde al día asignado para grabar la n-ésima toma tal y como aparece en la tabla de datos inicial.

Aplicando las restricciones de este problema, cada lista solución corresponderá a una permutación del conjunto $A$, definido en el apartado anterior.

## Según el modelo para el espacio de soluciones, ¿cuál es la función objetivo?¿Es un problema de maximización o minimización?

El modelo del espacio de soluciones consiste en una permutación del conjunto $A$ (teniendo en cuenta los datos de esta instancia en concreto). No obstante, esta permutación por sí misma no contiene información acerca de la bondad de la solución que representa.

Teniendo en cuenta que los actores cobran cada día que van al estudio, la función objetivo será aquella que calcule la cantidad de pagas totales que hay que realizar al finalizar toda la grabación. Dado que los actores cobran todos lo mismo, no es relevante el sueldo recibido, sino la cantidad de veces que se deberá entregar ese importe.

Sea: $n$ el número de tomas totales a grabar y $k$ el número máximo de tomas que se pueden grabar en un mismo día.

Teniendo en cuenta lo anterior, la cantidad de días óptima se definiría como: $\frac{n}{k}$. Teniendo en cuenta que, en un caso general, este número puede ser fraccionario, la parte decimal corresponderá siempre a un día adicional, por lo que definimos: $M = \lceil \frac{n}{k} \rceil$ como el número de días mínimo para grabar todas las tomas. En un caso general, la cantidad de permutaciones de días posibles se definiría como: $\frac{n!}{k!^M}$.

Cada toma en particular se define como: $T_{ij}$, representando la toma i-ésima asignada al día j.

Con todos estos datos, se puede definir la función objetivo como:

$$\sum_{j=1}^{M} count(T_{1j}| T_{2j}| ...| T_{kj})$$

El operador $|$ representa la función OR lógica. Debido a que cada toma se representa como un vector binario de actores, se puede aplicar la función OR sobre todas las tomas asignadas a un mismo día para determinar la cantidad de actores que **han entrado al estudio** ese día. La función $count()$ se encarga de contar esta cantidad de actores. Por último, solo queda sumar la cantidad de actores que entran al estudio cada uno de los días establecidos para la grabación.

La función, tal y como está planteada, es una función de **minimización**, pues el objetivo es minimizar la cantidad de veces que los actores entran en el estudio.

### Ejemplo
Para visualizar el funcionamiento de la función objetivo, consideremos un ejemplo sencillo. Supondremos que los datos son los mismos que ya tenemos, pero solo emplearemos las 6 primeras tomas. Consideremos también que solo se pueden grabar 3 tomas por día, dando un total de 2 días disponibles para realizar toda la grabación: $n = 6$, $k = 3$ y $M = 2$

Consideremos a continuación las soluciones: $[1,1,1,2,2,2]$ y $[1,2,1,2,1,2]$. ¿Cuál es mejor?

Primero, empleamos la lista solución para determinar qué tomas corresponden al día 1 y cuáles al 2. En el caso de la primera solución, las tomas 1, 2 y 3 de los datos se grabarán el día 1 y las tomas 4, 5 y 6, el día 2. Esto nos deja los siguientes conjuntos:

$$
\begin{equation}
D1 = \left( \begin{array}{rr} 1 & 1 & 1 & 1 & 1 & 0 & 0 & 0 & 0 & 0 \\ 0 & 0 & 1 & 1 & 1 & 0 & 0 & 0 & 0 & 0 \\ 0 & 1 & 0 & 0 & 1 & 0 & 1 & 0 & 0 & 0\end{array}\right)\hspace{1cm}
\end{equation}
$$

$$
\begin{equation}
D2 = \left( \begin{array}{rr} 1 & 1 & 0 & 0 & 0 & 0 & 1 & 1 & 0 & 0 \\ 0 & 1 & 0 & 1 & 0 & 0 & 0 & 1 & 0 & 0 \\ 1 & 1 & 0 & 1 & 1 & 0 & 0 & 0 & 0 & 0\end{array}\right)\hspace{1cm}
\end{equation}
$$

Con esta repartición de tomas en dos días, se calcula cuántos actores entran al estudio cada día. Para ello, se aplica la función OR lógica a **todas las tomas** de un día. Esto da como resultado:

$$
\begin{equation}
D1 = \left( \begin{array}{rr} 1 & 1 & 1 & 1 & 1 & 0 & 1 & 0 & 0 & 0\end{array}\right)\hspace{1cm}
|D1| = 6
\end{equation}
$$

$$
\begin{equation}
D2 = \left( \begin{array}{rr} 1 & 1 & 0 & 1 & 1 & 0 & 1 & 1 & 0 & 0\end{array}\right)\hspace{1cm}
|D2| = 6
\end{equation}
$$

Se puede ver que según esta solución ($[1,1,1,2,2,2]$), cada dían entran 6 actores al estudio, dando un total de $12$ pagas a abonar.

A continuación, consideremos la segunda solución ($[1,2,1,2,1,2]$). Según esta solución, las tomas 1, 3 y 5 se grabarán el día 1 y las tomas 2, 4 y 6 se grabarán el día 2. En este caso, las tomas repartidas por día quedarían:

$$
\begin{equation}
D1 = \left( \begin{array}{rr} 1 & 1 & 1 & 1 & 1 & 0 & 0 & 0 & 0 & 0 \\  0 & 1 & 0 & 0 & 1 & 0 & 1 & 0 & 0 & 0 \\ 0 & 1 & 0 & 1 & 0 & 0 & 0 & 1 & 0 & 0\end{array}\right)\hspace{1cm}
\end{equation}
$$

$$
\begin{equation}
D2 = \left( \begin{array}{rr} 0 & 0 & 1 & 1 & 1 & 0 & 0 & 0 & 0 & 0 \\ 1 & 1 & 0 & 0 & 0 & 0 & 1 & 1 & 0 & 0 \\ 1 & 1 & 0 & 1 & 1 & 0 & 0 & 0 & 0 & 0\end{array}\right)\hspace{1cm}
\end{equation}
$$

Aplicando la misma función OR lógica, obtenemos:

$$
\begin{equation}
D1 = \left( \begin{array}{rr} 1 & 1 & 1 & 1 & 1 & 0 & 1 & 1 & 0 & 0\end{array}\right)\hspace{1cm}
|D1| = 7
\end{equation}
$$

$$
\begin{equation}
D2 = \left( \begin{array}{rr} 1 & 1 & 1 & 1 & 1 & 0 & 1 & 1 & 0 & 0\end{array}\right)\hspace{1cm}
|D2| = 7
\end{equation}
$$

Por lo que el total de pagas a abonar según esta segunda solución es $14$. Esto convierte a la segunda solución en peor que la primera.

In [ ]:
# FUNCIÓN OBJETIVO IMPLEMENTADA
def calcular_pago(sol, data, verbose=True):
  try:
    assert len(sol) == len(data)
  except:
    print('ERROR: La cantidad de tomas en la solución no coincide con la cantidad de tomas en los datos')
    return float('inf')

  _data = data.copy()
  _data['Dia'] = pd.Series(sol) # Añadimos la solución como una nueva columna al dataframe # O(1)

  pago_total = 0
  for i in np.unique(sol):
    day_data = _data[_data['Dia'] == i] # Agrupamos las tomas por día asignado
    pago_total += np.sum(np.any(day_data.iloc[:,:-2], axis=0)) # Aplicamos la función OR y contamos la cantidad de actores

  if verbose: print(f'La cantidad de pagos a realizar según la solución {sol} es {pago_total}')
  return pago_total

calcular_pago([1,1,1,1,1,1,2,2,2,2,2,2,3,3,3,3,3,3,4,4,4,4,4,4,5,5,5,5,5,5], data);

La cantidad de pagos a realizar según la solución [1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5] es 38


La complejidad computacional de esta función de coste es: $3n^2 + \frac{3 + a}{k}n + 5$, siendo $n$ el número de tomas, $a$ el número de actores y $k$ las tomas máximas que se pueden grabar por día; lo que nos da una cota general del orden de $O(n^2)$.

# Solución por fuerza bruta

La solución por fuerza bruta para este problema es sencilla: generar todas las permutaciones posibles y probarlas todas hasta hallar la que minimiza la función objetivo.

Para esta solución, necesitamos una función capaz de calcular todas las permutaciones de una lista. A continuación, probamos cada permutación, calculando su coste y elegimos la permutación con el menor coste posible:

In [ ]:
def fuerza_bruta(data, k):
  # Calculamos parámetros de acuerdo a las fórmulas anteriores
  n = len(data)
  M = math.ceil(n/k)
  print(f'{len(data.loc[0])-1} Actores\n{n} Tomas\nMáximo {k} tomas por día, se requiere un mínimo de {M} días\n')

  A = [x for x in range(1, M+1) for _ in range(k)] # Generamos la solución base [1,1,1,1,1,1,2,2,2,2,2,2,3,3,3,3,3,3,4,4,4,4,4,4,5,5,5,5,5,5] para n=30 y k=6
  A = A[:n]
  permutaciones = itertools.permutations(A)

  coste_minimo = float('inf')
  sol_optima = []
  for p in permutaciones:
    coste_actual = calcular_pago(p, data)
    if coste_actual < coste_minimo:
      coste_minimo = coste_actual
      sol_optima = p

  print(f'La mejor solución encontrada es {sol_optima}, con un coste de {coste_minimo}')
  return sol_optima, coste_minimo

Debido a que esta solución debe calcular todas las permutaciones posibles, se ha establecido con anterioridad que esta operación genera $\frac{n!}{k!^M}$ soluciones, siendo $n$ la cantidad de tomas. Esto demuestra que la solución por fuerza bruta va a tener una complejidad del orden $O(n!)$, una complejidad totalmente inviable.

# Solución por Búsqueda Aleatoria
Una vez comprobada la inefectividad del algoritmo por fuerza bruta, es el momento de explorar otras alternativas. La primera alternativa, y la más sencilla es la Búsqueda Aleatoria. Como su nombre indica, este algoritmo se basa en generar una cantidad $C$ de soluciones aleatorias, de las cuales se elegirá la mejor. A diferencia del algortimo de fuerza bruta, es probable que no se alcance la solución óptima, ni siquiera una solución buena, pero a cambio es un algoritmo mucho más rápido debido a que sólo se exploran $C$ soluciones.

In [ ]:
def busqueda_aleatoria(C, data, k, verbose=True):
  # Calculamos los parámetros de acuerdo a las fórmulas anteriores
  n = len(data)
  M = math.ceil(n/k)
  if verbose: print(f'{len(data.loc[0])-1} Actores\n{n} Tomas\nMáximo {k} tomas por día, se requiere un mínimo de {M} días\n')

  A = [x for x in range(1, M+1) for _ in range(k)] # Solución base
  A = A[:n]

  coste_minimo = float('inf')
  sol_optima = []
  for _ in range(C):
    random.shuffle(A) # Permutación aleatoria de la solución base
    coste_actual = calcular_pago(A, data, verbose=False)

    if coste_actual < coste_minimo:
      coste_minimo = coste_actual
      sol_optima = A
      if verbose: print(f'Encontrada mejor solución con coste {coste_minimo}')

  if verbose: print(f'La mejor solución encontrada es {sol_optima}, con un coste de {coste_minimo}')
  return sol_optima, coste_minimo

In [ ]:
busqueda_aleatoria(10000, data, 6);

10 Actores
30 Tomas
Máximo 6 tomas por día, se requiere un mínimo de 5 días

Encontrada mejor solución con coste 37
Encontrada mejor solución con coste 36
Encontrada mejor solución con coste 35
Encontrada mejor solución con coste 34
Encontrada mejor solución con coste 32
La mejor solución encontrada es [5, 3, 5, 1, 3, 1, 1, 2, 1, 2, 1, 4, 4, 3, 1, 4, 2, 3, 2, 5, 4, 2, 3, 5, 2, 5, 4, 4, 5, 3], con un coste de 32


Esta primera solución tiene una complejidad de: $Cn^2 + (C + 1)n + 4C + 8$, lo que nos da una cota superior de $O(n^2)$ con constante multiplicativa $C$, siendo $C$ el número de soluciones aleatorias a comprobar.

# Búsqueda Local
Tras considerar la búsqueda aleatoria, el siguiente paso es intensificar las soluciones obtenidas. En este ámbito, se van a emplear una serie de operadores para obtener soluciones vecinas prometedoras dada una solución inicial. El objetivo es generar todos los posibles vecinos y seleccionar el que mejore más la solución inicial.

## Primer operador de generación de vecinos: 2-opt
En este primer operador, se intercambian 2 nodos de la solución inicial. En este caso, se comprueban todos los posibles intercambios, seleccionando como vecino el mejor de ellos. Este operador tiene complejidad $O(n^4)$, $O(n^2)$ si no se tiene en cuenta la complejidad de la función de coste.

In [ ]:
def _2_opt(solucion, data, verbose=True):
  mejor_vecino = []
  coste_minimo = float('inf')

  for i in range(1, len(solucion)-1):
    for j in range(i+1, len(solucion)):
      vecino = solucion[:i] + [solucion[j]] + solucion[i+1:j] + [solucion[i]] + solucion[j+1:]
      coste_actual = calcular_pago(vecino, data, verbose=False) # O(n^2)

      if coste_actual < coste_minimo:
        coste_minimo = coste_actual
        mejor_vecino = vecino

  if verbose: print(f'Vecino prometedor con coste {coste_minimo} encontrado: {mejor_vecino}')
  return mejor_vecino

## Segundo operador de generación de vecinos: inserción
Este segundo operador extrae un elemento en una posición de la solución y la inserta en otra distinta. En este caso, se comprueban todas las posibles inserciones, quedándonos con la mejor. Este operador también tiene complejidad $O(n^4)$, $O(n^2)$ si no se tiene en cuenta la complejidad de la función de coste.

In [ ]:
def _insertion(solucion, data, verbose=True):
  mejor_vecino = []
  coste_minimo = float('inf')

  for i in range(len(solucion)):
    vecino = solucion.copy()
    elemento = vecino.pop(i) # Extracción
    for j in range(len(solucion)):
      if i != j:
        vecino.insert(j, elemento) # Inserción

        coste_actual = calcular_pago(vecino, data, verbose=False) # O(n^2)

        if coste_actual < coste_minimo:
          coste_minimo = coste_actual
          mejor_vecino = vecino.copy()

        vecino.pop(j)

  if verbose: print(f'Vecino prometedor con coste {coste_minimo} encontrado: {mejor_vecino}')
  return mejor_vecino

## Algoritmo de búsqueda local
Una vez definidos los diferentes operadores de generación de soluciones vecinas, se procede a implementar el algoritmo de búsqueda local, que emplea estos operadores para hallar el mejor vecino dada una solución inicial aleatoria. En este caso concreto, se generan vecinos óptimos hasta que la solución deja de mejorar. Si una solución no tiene un vecino mejor, significa que se ha alcanzado un mínimo local, por lo que se detiene la búsqueda.

In [ ]:
def busqueda_local(data, k, func_generar_vecino=_2_opt, verbose=True):
  # Calculamos los parámetros de acuerdo a las fórmulas anteriores
  n = len(data)
  M = math.ceil(n/k)
  if verbose: print(f'{len(data.loc[0])-1} Actores\n{n} Tomas\nMáximo {k} tomas por día, se requiere un mínimo de {M} días\n')

  # Generamos una solución inicial aleatoria
  A = [x for x in range(1, M+1) for _ in range(k)] # Solución base
  A = A[:n]
  random.shuffle(A)
  if verbose: print(f'Solución inicial aleatoria: {A}')

  coste_minimo = calcular_pago(A, data, verbose=False)
  sol_optima = A

  # Continuaremos generando vecinos hasta que las soluciones no mejoren. Esto indicará que hemos alcanzado un mínimo local
  iteraciones = 0
  while(1):
    iteraciones += 1

    vecino = func_generar_vecino(sol_optima, data, verbose=verbose)
    coste_actual = calcular_pago(vecino, data, verbose=False)

    if coste_actual < coste_minimo:
      coste_minimo = coste_actual
    else:
      if verbose: print(f'Encontrado mínimo local en la iteración {iteraciones}, con coste {coste_minimo}: {sol_optima}')
      return sol_optima, coste_minimo

    sol_optima = vecino

In [ ]:
busqueda_local(data, 6, _2_opt);

10 Actores
30 Tomas
Máximo 6 tomas por día, se requiere un mínimo de 5 días

Solución inicial aleatoria: [5, 1, 1, 4, 4, 2, 5, 4, 4, 2, 3, 1, 2, 1, 4, 4, 2, 3, 3, 5, 1, 5, 2, 3, 5, 3, 2, 3, 1, 5]
Vecino prometedor con coste 33 encontrado: [5, 1, 1, 4, 4, 2, 5, 4, 4, 3, 3, 1, 2, 1, 4, 4, 2, 3, 3, 5, 1, 5, 2, 3, 5, 3, 2, 2, 1, 5]
Vecino prometedor con coste 32 encontrado: [5, 2, 1, 4, 4, 1, 5, 4, 4, 3, 3, 1, 2, 1, 4, 4, 2, 3, 3, 5, 1, 5, 2, 3, 5, 3, 2, 2, 1, 5]
Vecino prometedor con coste 31 encontrado: [5, 2, 4, 4, 4, 1, 5, 1, 4, 3, 3, 1, 2, 1, 4, 4, 2, 3, 3, 5, 1, 5, 2, 3, 5, 3, 2, 2, 1, 5]
Vecino prometedor con coste 30 encontrado: [5, 2, 5, 4, 4, 1, 5, 1, 4, 3, 3, 1, 2, 1, 4, 4, 2, 3, 3, 5, 1, 5, 2, 3, 4, 3, 2, 2, 1, 5]
Vecino prometedor con coste 29 encontrado: [5, 2, 5, 4, 4, 1, 5, 3, 4, 3, 1, 1, 2, 1, 4, 4, 2, 3, 3, 5, 1, 5, 2, 3, 4, 3, 2, 2, 1, 5]
Vecino prometedor con coste 29 encontrado: [5, 2, 5, 4, 4, 1, 5, 3, 4, 3, 1, 1, 2, 1, 4, 4, 2, 3, 3, 5, 1, 5, 2, 3, 4, 3, 2, 2, 1, 5]


In [ ]:
busqueda_local(data, 6, _insertion);

10 Actores
30 Tomas
Máximo 6 tomas por día, se requiere un mínimo de 5 días

Solución inicial aleatoria: [2, 1, 2, 5, 4, 1, 3, 5, 3, 2, 5, 3, 5, 4, 1, 4, 3, 1, 3, 2, 4, 1, 4, 3, 5, 2, 5, 2, 4, 1]
Vecino prometedor con coste 36 encontrado: [1, 2, 5, 4, 1, 3, 5, 3, 2, 5, 3, 5, 4, 1, 4, 3, 1, 3, 2, 4, 1, 4, 3, 5, 2, 5, 2, 4, 1, 2]
Vecino prometedor con coste 34 encontrado: [2, 5, 4, 1, 3, 5, 3, 1, 2, 5, 3, 5, 4, 1, 4, 3, 1, 3, 2, 4, 1, 4, 3, 5, 2, 5, 2, 4, 1, 2]
Vecino prometedor con coste 32 encontrado: [2, 5, 4, 1, 3, 5, 3, 1, 2, 5, 3, 5, 4, 1, 4, 2, 3, 1, 3, 4, 1, 4, 3, 5, 2, 5, 2, 4, 1, 2]
Vecino prometedor con coste 32 encontrado: [5, 2, 4, 1, 3, 5, 3, 1, 2, 5, 3, 5, 4, 1, 4, 2, 3, 1, 3, 4, 1, 4, 3, 5, 2, 5, 2, 4, 1, 2]
Encontrado mínimo local en la iteración 4, con coste 32: [2, 5, 4, 1, 3, 5, 3, 1, 2, 5, 3, 5, 4, 1, 4, 2, 3, 1, 3, 4, 1, 4, 3, 5, 2, 5, 2, 4, 1, 2]


En general, el algoritmo de búsqueda local tiene una complejidad de $In^4 + (I + 1)n^2 + 2n + 5I + 7$, siendo $I$ el número de iteraciones que realiza el bucle *while*. Esto nos da una complejidad de $O(n^4)$ con constante multiplicativa $I$.

## Multiarranque
El problema principal de los algoritmos de búsqueda local es la posibilidad de atascarse en mínimos locales. Esto se puede solucionar lanzando el algoritmo varias veces, con soluciones iniciales diferentes. Al empezar a intensificar desde distintos puntos, es probable que algunas de las soluciones lleguen al mínimo global.

En este caso, el multiarranque consiste en un bucle que llama $i$ veces al algoritmo de búsqueda local. Cada vez que se llama, se genera una solución inicial aleatoria, por lo que la intensificación se produce desde un punto distinto. Como en la mayoría de estas funciones, guardamos la mejor solución global.

La complejidad de este algoritmo es muy sencilla, puesto que consiste en repetir el algoritmo de búsqueda local una cantidad $A$ de veces. Esto nos da una complejidad de $O(An^4)$.

In [ ]:
def multiarranque(data, k, iteraciones=5, func_generar_vecino=_2_opt, verbose=True):
  # Ejecutamos varias veces un mismo algoritmo de búsqueda local, con diferentes soluciones iniciales aleatorias
  sol_optima = []
  coste_minimo = float('inf')

  for i in range(iteraciones):
    if verbose: print(f'ITERACIÓN {i+1}')
    sol_actual, _ = busqueda_local(data, k, func_generar_vecino, verbose=False)
    coste_actual = calcular_pago(sol_actual, data, verbose=False)

    if coste_actual < coste_minimo:
      coste_minimo = coste_actual
      sol_optima = sol_actual
      if verbose:
        print(f'MEJOR SOLUCIÓN HASTA EL MOMENTO: {sol_optima}')
        print(f'CON COSTE {coste_minimo}\n')
    else:
      if verbose: print('SOLUCIÓN INFRUCTUOSA. NO SE HA ALCANZADO MEJORÍA\n')

  if verbose:
    print(f'MEJOR SOLUCIÓN GLOBAL: {sol_optima}')
    print(f'CON COSTE: {coste_minimo}')
  return sol_optima, coste_minimo


In [ ]:
multiarranque(data, 6, iteraciones=6, func_generar_vecino=_2_opt);

ITERACIÓN 1
MEJOR SOLUCIÓN HASTA EL MOMENTO: [3, 2, 5, 5, 4, 3, 3, 4, 4, 1, 5, 1, 2, 1, 5, 3, 5, 2, 5, 2, 4, 1, 3, 2, 3, 1, 1, 4, 2, 4]
CON COSTE 29

ITERACIÓN 2
SOLUCIÓN INFRUCTUOSA. NO SE HA ALCANZADO MEJORÍA

ITERACIÓN 3
SOLUCIÓN INFRUCTUOSA. NO SE HA ALCANZADO MEJORÍA

ITERACIÓN 4
SOLUCIÓN INFRUCTUOSA. NO SE HA ALCANZADO MEJORÍA

ITERACIÓN 5
MEJOR SOLUCIÓN HASTA EL MOMENTO: [2, 4, 3, 3, 3, 5, 5, 2, 5, 2, 3, 2, 4, 1, 3, 5, 1, 1, 1, 4, 1, 3, 4, 1, 5, 2, 5, 4, 2, 4]
CON COSTE 27

ITERACIÓN 6
SOLUCIÓN INFRUCTUOSA. NO SE HA ALCANZADO MEJORÍA

MEJOR SOLUCIÓN GLOBAL: [2, 4, 3, 3, 3, 5, 5, 2, 5, 2, 3, 2, 4, 1, 3, 5, 1, 1, 1, 4, 1, 3, 4, 1, 5, 2, 5, 4, 2, 4]
CON COSTE: 27


In [ ]:
multiarranque(data, 6, iteraciones=6, func_generar_vecino=_insertion);

ITERACIÓN 1
MEJOR SOLUCIÓN HASTA EL MOMENTO: [5, 2, 1, 3, 3, 1, 1, 4, 4, 2, 3, 2, 3, 1, 3, 2, 4, 4, 5, 1, 4, 5, 3, 5, 2, 2, 1, 5, 5, 4]
CON COSTE 34

ITERACIÓN 2
MEJOR SOLUCIÓN HASTA EL MOMENTO: [4, 5, 4, 4, 4, 3, 3, 1, 5, 5, 4, 1, 1, 2, 4, 1, 2, 2, 5, 3, 2, 1, 2, 2, 1, 5, 3, 3, 5, 3]
CON COSTE 30

ITERACIÓN 3
SOLUCIÓN INFRUCTUOSA. NO SE HA ALCANZADO MEJORÍA

ITERACIÓN 4
SOLUCIÓN INFRUCTUOSA. NO SE HA ALCANZADO MEJORÍA

ITERACIÓN 5
SOLUCIÓN INFRUCTUOSA. NO SE HA ALCANZADO MEJORÍA

ITERACIÓN 6
SOLUCIÓN INFRUCTUOSA. NO SE HA ALCANZADO MEJORÍA

MEJOR SOLUCIÓN GLOBAL: [4, 5, 4, 4, 4, 3, 3, 1, 5, 5, 4, 1, 1, 2, 4, 1, 2, 2, 5, 3, 2, 1, 2, 2, 1, 5, 3, 3, 5, 3]
CON COSTE: 30


Tras comprobar el funcionamiento de ambos operadores de generación de vecinos, el primero de ellos, 2-opt, ha resultado ser el más prometedor, no solo en cuanto a rendimiento, sino en cuanto a la calidad de los vecinos generados. Dada la naturaleza del problema, resulta más eficiente intercambiar los horarios de 2 tomas, sin afectar al resto, que insertar una toma desplazando todas las demás que se graban ese día. Ese desplazamiento causa que los actores tengan que acudir al estudio un día más, causando soluciones menos efectivas.

# Algoritmo Voraz
La última opción explorada aquí es un algoritmo voraz. Con el criterio voraz adecuado, este algoritmo tiene el potencial de ser el más rápido y eficiente. En este caso, se sabe que los actores cobran por cada día que van al estudio, por lo que agrupar las tomas en base a cuántos actores comparten debería minimizar la cantidad de desplazamientos.

En este caso, el primer paso del algoritmo es localizar la toma con mayor número de actores en ella y asignarle el día 1. A continuación, se asigna un valor de proximidad (cantidad de actores en común) a cada una de las otras tomas. Las tomas se ordenan en base a este valor de proximidad y se asignan en base a cuán proximas son con respecto a la primera. Una vez que se han asignado las $k$ tomas de un día, el proceso se repite para los $M$ días.

In [ ]:
def algoritmo_voraz(data, k, verbose=True):
  # Copiamos los datos para no modificar los originales
  data_greedy = data.copy()

  # Calculamos los parámetros de acuerdo a las fórmulas anteriores
  n = len(data_greedy)
  M = math.ceil(n/k)
  if verbose: print(f'{len(data_greedy.loc[0])-1} Actores\n{n} Tomas\nMáximo {k} tomas por día, se requiere un mínimo de {M} días\n')

  # Generamos una solución vacía
  sol = np.zeros(n)

  for d in range(1, M+1):
    # Generamos un vector para almacenar los valores de proximidad
    prox = np.zeros(len(data_greedy))

    # Asignamos el día 1 a la toma con más actores
    max_index = data_greedy['Total'].idxmax()
    sol[max_index] = d
    prox[data_greedy.index.get_loc(max_index)] = -999

    # Generamos los valores de proximidad
    for j, i in enumerate(data_greedy.index):
      if i != max_index:
        prox[j] = np.sum(np.all(data_greedy.loc[[i, max_index], list(data_greedy.columns[:-1])], axis=0))

    # Añadimos los velores de proximidad y ordenamos los datos en torno a estos
    data_greedy['Prox'] = prox
    sorted_data = data_greedy.sort_values('Prox', ascending=False)

    # Escogemos las k-1 tomas más próximas a la primera y les asignamos el mismo día
    mayor_prox_indices = sorted_data[:min(k-1, len(sorted_data))].index
    sol[mayor_prox_indices] = d

    # Eliminamos las tomas ya asignadas para continuar con el siguiente día
    data_greedy.drop(mayor_prox_indices, axis=0, inplace=True)
    data_greedy.drop(max_index, axis=0, inplace=True)

  sol = list(sol)
  coste = calcular_pago(sol, data, verbose=False)
  if verbose: print(f'Encontrada solución con coste {coste}: {sol}')
  return sol, coste

In [ ]:
algoritmo_voraz(data, 6);

10 Actores
30 Tomas
Máximo 6 tomas por día, se requiere un mínimo de 5 días

Encontrada solución [1.0, 4.0, 3.0, 3.0, 3.0, 1.0, 1.0, 2.0, 2.0, 2.0, 1.0, 2.0, 4.0, 2.0, 3.0, 5.0, 4.0, 4.0, 3.0, 1.0, 5.0, 1.0, 4.0, 5.0, 2.0, 4.0, 5.0, 5.0, 3.0, 5.0] con coste 34


Como se puede observar, el algoritmo voraz propone soluciones peores que el algoritmo de búsqueda local con multiarranque. No obstante, es bastante más rápido, por lo que resulta más idóneo cuando el tamaño del problema crece considerablemente. La complejidad de este algoritmo es $\frac{1}{k}n^3 + (2 + \frac{9+log(n)+a}{k})n^2 + (1 + \frac{10}{k})n + 5$, es decir, $O(n^3)$, una de las más bajas consideradas. Esto, sumado a la optimización que proporciona la librería *pandas*, lo convierten en el algoritmo más rápido de los aquí considerados.

# Algoritmo Voraz + Búsqueda Local
Para paliar la baja calidad de las soluciones del algoritmo voraz, se va a combinar este con el algoritmo de búsqueda local. El resultado debería de ser todavía relativamente rápido, al tiempo que genera soluciones mejores. Dado que la complejidad del algoritmo de búsqueda local es $O(n^4)$, este algoritmo comparte la misma cota.

In [ ]:
def algoritmo_voraz_busqueda_local(data, k, func_generar_vecino=_2_opt, verbose=True):
  # Calculamos los parámetros de acuerdo a las fórmulas anteriores
  n = len(data)
  M = math.ceil(n/k)
  if verbose: print(f'{len(data.loc[0])-1} Actores\n{n} Tomas\nMáximo {k} tomas por día, se requiere un mínimo de {M} días\n')

  # En esta versión, la solución inicial viene dada por el algoritmo voraz
  A, _ = algoritmo_voraz(data, k, verbose=False)
  if verbose: print(f'Solución inicial propuesta por el algoritmo voraz: {A}')

  coste_minimo = calcular_pago(A, data, verbose=False)
  sol_optima = A

  # Continuaremos generando vecinos hasta que las soluciones no mejoren. Esto indicará que hemos alcanzado un mínimo local
  iteraciones = 0
  while(1):
    iteraciones += 1

    vecino = func_generar_vecino(sol_optima, data, verbose=verbose)
    coste_actual = calcular_pago(vecino, data, verbose=False)

    if coste_actual < coste_minimo:
      coste_minimo = coste_actual
    else:
      if verbose: print(f'Encontrado mínimo local en la iteración {iteraciones}, con coste {coste_minimo}: {sol_optima}')
      return sol_optima, coste_minimo

    sol_optima = vecino

In [ ]:
algoritmo_voraz_busqueda_local(data, 6);

10 Actores
30 Tomas
Máximo 6 tomas por día, se requiere un mínimo de 5 días

Solución inicial propuesta por el algoritmo voraz: [1.0, 4.0, 3.0, 3.0, 3.0, 1.0, 1.0, 2.0, 2.0, 2.0, 1.0, 2.0, 4.0, 2.0, 3.0, 5.0, 4.0, 4.0, 3.0, 1.0, 5.0, 1.0, 4.0, 5.0, 2.0, 4.0, 5.0, 5.0, 3.0, 5.0]
Vecino prometedor con coste 32 encontrado: [1.0, 4.0, 3.0, 3.0, 3.0, 1.0, 1.0, 2.0, 2.0, 2.0, 1.0, 2.0, 4.0, 2.0, 3.0, 5.0, 4.0, 5.0, 3.0, 1.0, 5.0, 1.0, 4.0, 5.0, 2.0, 4.0, 4.0, 5.0, 3.0, 5.0]
Vecino prometedor con coste 30 encontrado: [1.0, 4.0, 3.0, 3.0, 3.0, 1.0, 1.0, 2.0, 2.0, 2.0, 1.0, 2.0, 4.0, 2.0, 3.0, 5.0, 4.0, 5.0, 5.0, 1.0, 3.0, 1.0, 4.0, 5.0, 2.0, 4.0, 4.0, 5.0, 3.0, 5.0]
Vecino prometedor con coste 29 encontrado: [1.0, 4.0, 3.0, 3.0, 3.0, 1.0, 1.0, 2.0, 2.0, 2.0, 1.0, 2.0, 4.0, 5.0, 3.0, 2.0, 4.0, 5.0, 5.0, 1.0, 3.0, 1.0, 4.0, 5.0, 2.0, 4.0, 4.0, 5.0, 3.0, 5.0]
Vecino prometedor con coste 29 encontrado: [1.0, 4.0, 3.0, 3.0, 3.0, 1.0, 1.0, 2.0, 2.0, 2.0, 1.0, 2.0, 4.0, 5.0, 3.0, 2.0, 4.0, 5.0, 5.0, 

# Comparativa de tiempos entre los diferentes algoritmos
Una vez planteados todos los algoritmos, se comparan sus resultados en una tabla para determinar cuál de ellos proporciona el mejor rendimiento. Para ello, se mide el tiempo que tarda el algoritmo en alcanzar una solución y se compara la bondad de dicha solución.

*Nota: Debido a la aleatoriedad implícita de la mayoría de métodos, es posible que los tiempos y soluciones cambien de una iteración a otra.*

In [ ]:
start = time.time()
_, coste = busqueda_aleatoria(10000, data, 6, verbose=False)
print(f'Búsqueda Aleatoria: obtenido coste {coste} en {round(time.time() - start, 2)} segundos')

start = time.time()
_, coste = busqueda_local(data, 6, _2_opt, verbose=False)
print(f'Búsqueda Local con selección de vecino mediante 2-opt: obtenido coste {coste} en {round(time.time() - start, 2)} segundos')

start = time.time()
_, coste = busqueda_local(data, 6, _insertion, verbose=False)
print(f'Búsqueda Local con selección de vecino mediante Insertion: obtenido coste {coste} en {round(time.time() - start, 2)} segundos')

start = time.time()
_, coste = multiarranque(data, 6, func_generar_vecino=_2_opt, verbose=False)
print(f'Multiarranque con selección de vecino mediante 2-opt y 5 iteraciones: obtenido coste {coste} en {round(time.time() - start, 2)} segundos')

start = time.time()
_, coste = algoritmo_voraz(data, 6, verbose=False)
print(f'Algoritmo Voraz: obtenido coste {coste} en {round(time.time() - start, 2)} segundos')

start = time.time()
_, coste = algoritmo_voraz_busqueda_local(data, 6, _2_opt, verbose=False)
print(f'Algoritmo Voraz + Búsqueda Local con selección de vecino mediante 2-opt: obtenido coste {coste} en {round(time.time() - start, 2)} segundos')

Búsqueda Aleatoria: obtenido coste 31 en 60.29 segundos
Búsqueda Local con selección de vecino mediante 2-opt: obtenido coste 33 en 12.19 segundos
Búsqueda Local con selección de vecino mediante Insertion: obtenido coste 32 en 25.95 segundos
Multiarranque con selección de vecino mediante 2-opt y 5 iteraciones: obtenido coste 28 en 82.26 segundos
Algoritmo Voraz: obtenido coste 34 en 0.2 segundos
Algoritmo Voraz + Búsqueda Local con selección de vecino mediante 2-opt: obtenido coste 29 en 10.34 segundos


|            **Modelo**            | **Solución Obtenida** | **Tiempo (s)** | **Complejidad** |
|:--------------------------------:|:---------------------:|:--------------:|:---------------:|
|        Búsqueda Aleatoria        |           31          |      60.29     |        $O(Cn^2)$         |
|      Búsqueda Local (2-opt)      |           33          |      12.19     |        $O(In^4)$         |
|    Búsqueda Local (Insertion)    |           32          |      25.95     |         $O(In^4)$        |
|       Multiarranque (2-opt)      |           28          |      82.26     |       $O(An^4)$          |
|          Algoritmo Voraz         |           34          |       0.2      |        $O(n^3)$         |
| Algoritmo Voraz + Búsqueda Local |           29          |      10.34     |       $O(In^4)$          |

# Conclusiones

Como se puede observar en la tabla, muchos de los algoritmos planteados tienen complejidades similares en el caso promedio. No obstante, algunos algoritmos tienen constantes multiplicativas muy grandes que los hacen más lentos, o se basan en llamar a otros algoritmos un número determinado de veces, arrastrando todas las constantes multiplicativas de los algoritmos a los que llaman y resultando en un mayor gasto de tiempo (por ejemplo, este es el caso del algoritmo de *multiarranque*).

En cuanto a rendimiento como tal, se puede observar que el algoritmo de *multiarranque* es el más lento de todos, pero también es el que mejores soluciones encuentra. Invirtiendo más tiempo, es posible que este algoritmo proporcione muy buenas soluciones. No obstante, es el más lento de todos.

En el otro extremo está el algoritmo voraz. Un algoritmo increíblemente rápido, que proporciona buenas soluciones aunque no tan buenas como los demás algoritmos. Es una buena opción si se precisa encontrar cualquier solución en un tiempo récord, pero no funciona demasiado bien si se pretende encontrar la solución óptima (o cercana al óptimo).

En un punto medio se encuentra el último algoritmo analizado, la combinación del algoritmo voraz más la sucesiva búsqueda local. Este algoritmo encuentra un agradable punto medio, donde sus soluciones son de mayor calidad que el algoritmo voraz, estando cerca de las soluciones propuestas por el algoritmo de *multiarranque*, sin embargo el tiempo invertido es bastante aceptable.

Cabe destacar que tanto el algoritmo voraz como el voraz + búsqueda local son totalmente deterministas, por lo que no pueden diversificar en busca de mejores soluciones. El algoritmo de *multiarranque*, con su gran componente estocástica, es el que más diversifica de todos, de ahí su capacidad para encontrar buenas soluciones, sin embargo la diversificación requiere una gran inversión de tiempo.

# Juegos de Entrada Aleatorios

Al plantear un algoritmo como solución de un problema concreto, se corre el riesgo de especializarlo demasiado. Esto supone que el algoritmo falla a la hora de adaptarse a otros problemas, o incluso a planteamientos diferentes del mismo problema. Para evitar esto, se van a generar varios juegos de datos de entrada aleatorios para evaluar cómo se adapta el algoritmo a ellos.

In [ ]:
# Generador de datos aleatorios
def generar_datos(n=None, a=None, k=None, verbose=True):
  """
  n: Cantidad de tomas a grabar
  a: Cantidad de actores participantes
  k: Número máximo de tomas por día
  """

  if n == None:
    n = random.randint(15, 100)
  if a == None:
    a = random.randint(3, 30)
  if k == None:
    k = random.randint(2, 10)

  initial_config = np.rint(np.random.rand(n, a))

  data = pd.DataFrame(initial_config, index=np.arange(n), columns=['Actor ' + str(x) for x in np.arange(1, a+1)])
  data.index.name = 'Toma'
  data['Total'] = np.sum(data, axis=1)
  if verbose: display(data)

  return data, k

In [ ]:
d, k = generar_datos(n=63, a=15, k=7);

Actor 1  Actor 2  Actor 3  Actor 4  Actor 5  Actor 6  Actor 7  Actor 8  \
Toma                                                                           
0         1.0      0.0      0.0      1.0      1.0      1.0      0.0      1.0   
1         0.0      1.0      1.0      1.0      1.0      1.0      1.0      0.0   
2         0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0   
3         1.0      0.0      0.0      0.0      0.0      0.0      1.0      0.0   
4         1.0      1.0      0.0      1.0      1.0      0.0      0.0      1.0   
...       ...      ...      ...      ...      ...      ...      ...      ...   
58        0.0      1.0      1.0      1.0      1.0      1.0      0.0      1.0   
59        0.0      0.0      0.0      0.0      0.0      1.0      1.0      0.0   
60        1.0      1.0      1.0      1.0      1.0      0.0      1.0      1.0   
61        0.0      1.0      0.0      0.0      0.0      1.0      0.0      0.0   
62        0.0      0.0      0.0      1.0      1.0      1.0      1.0      0.0   

      Actor 9  Actor 10  Actor 11  Actor 12  Actor 13  Actor 14  Actor 15  \
Toma                                                                        
0         1.0       1.0       0.0       1.0       0.0       0.0       1.0   
1         0.0       0.0       0.0       1.0       1.0       1.0       0.0   
2         1.0       1.0       1.0       0.0       1.0       1.0       0.0   
3         0.0       0.0       0.0       0.0       0.0       0.0       0.0   
4         0.0       1.0       0.0       1.0       0.0       1.0       1.0   
...       ...       ...       ...       ...       ...       ...       ...   
58        1.0       0.0       1.0       0.0       0.0       0.0       1.0   
59        1.0       0.0       0.0       1.0       1.0       0.0       1.0   
60        0.0       1.0       0.0       1.0       0.0       1.0       1.0   
61        0.0       1.0       1.0       1.0       0.0       1.0       0.0   
62        0.0       1.0       0.0       0.0       0.0       0.0       0.0   

      Total  
Toma         
0       9.0  
1       9.0  
2       6.0  
3       2.0  
4       9.0  
...     ...  
58      9.0  
59      6.0  
60     11.0  
61      6.0  
62      5.0  

[63 rows x 16 columns]

In [ ]:
busqueda_aleatoria(10000, d, k);

15 Actores
63 Tomas
Máximo 7 tomas por día, se requiere un mínimo de 9 días

Encontrada mejor solución con coste 134
Encontrada mejor solución con coste 133
Encontrada mejor solución con coste 132
Encontrada mejor solución con coste 131
Encontrada mejor solución con coste 130
La mejor solución encontrada es [8, 1, 1, 3, 3, 4, 8, 8, 2, 2, 6, 1, 1, 5, 9, 9, 3, 2, 4, 5, 1, 9, 6, 9, 2, 8, 6, 9, 2, 5, 9, 5, 8, 5, 9, 4, 2, 3, 4, 5, 8, 7, 6, 3, 3, 1, 5, 7, 2, 7, 6, 8, 7, 7, 7, 6, 7, 4, 6, 3, 4, 4, 1], con un coste de 130


In [ ]:
busqueda_local(d, k, func_generar_vecino=_2_opt);

15 Actores
63 Tomas
Máximo 7 tomas por día, se requiere un mínimo de 9 días

Solución inicial aleatoria: [1, 9, 9, 7, 3, 7, 3, 8, 5, 8, 6, 4, 9, 3, 8, 4, 1, 9, 4, 6, 3, 6, 2, 8, 7, 3, 8, 4, 2, 5, 9, 9, 6, 1, 2, 4, 7, 3, 1, 1, 2, 1, 2, 1, 5, 8, 6, 5, 6, 9, 7, 5, 8, 2, 6, 5, 5, 7, 2, 3, 4, 7, 4]
Vecino prometedor con coste 131 encontrado: [1, 9, 9, 7, 3, 2, 3, 8, 5, 8, 6, 4, 9, 3, 8, 4, 1, 9, 4, 6, 3, 6, 2, 8, 7, 3, 8, 4, 2, 5, 9, 9, 6, 1, 2, 4, 7, 3, 1, 1, 2, 1, 2, 1, 5, 8, 6, 5, 6, 9, 7, 5, 8, 2, 6, 5, 5, 7, 7, 3, 4, 7, 4]
Vecino prometedor con coste 130 encontrado: [1, 9, 9, 1, 3, 2, 3, 8, 5, 8, 6, 4, 9, 3, 8, 4, 1, 9, 4, 6, 3, 6, 2, 8, 7, 3, 8, 4, 2, 5, 9, 9, 6, 1, 2, 4, 7, 3, 1, 7, 2, 1, 2, 1, 5, 8, 6, 5, 6, 9, 7, 5, 8, 2, 6, 5, 5, 7, 7, 3, 4, 7, 4]
Vecino prometedor con coste 128 encontrado: [1, 9, 9, 1, 3, 2, 3, 1, 5, 8, 6, 4, 9, 3, 8, 4, 8, 9, 4, 6, 3, 6, 2, 8, 7, 3, 8, 4, 2, 5, 9, 9, 6, 1, 2, 4, 7, 3, 1, 7, 2, 1, 2, 1, 5, 8, 6, 5, 6, 9, 7, 5, 8, 2, 6, 5, 5, 7, 7, 3, 4, 7, 4]
Vec

In [ ]:
algoritmo_voraz(d, k);

15 Actores
63 Tomas
Máximo 7 tomas por día, se requiere un mínimo de 9 días

Encontrada solución [3.0, 2.0, 8.0, 9.0, 1.0, 9.0, 5.0, 8.0, 5.0, 5.0, 2.0, 8.0, 3.0, 2.0, 8.0, 6.0, 2.0, 9.0, 1.0, 4.0, 2.0, 7.0, 4.0, 1.0, 9.0, 6.0, 7.0, 4.0, 7.0, 4.0, 9.0, 5.0, 6.0, 4.0, 9.0, 1.0, 3.0, 5.0, 8.0, 6.0, 7.0, 1.0, 6.0, 7.0, 7.0, 3.0, 7.0, 6.0, 2.0, 1.0, 5.0, 4.0, 3.0, 6.0, 3.0, 3.0, 4.0, 5.0, 2.0, 8.0, 1.0, 8.0, 9.0] con coste 130


In [ ]:
algoritmo_voraz_busqueda_local(d, k);

15 Actores
63 Tomas
Máximo 7 tomas por día, se requiere un mínimo de 9 días

Solución inicial propuesta por el algoritmo voraz: [3.0, 2.0, 8.0, 9.0, 1.0, 9.0, 5.0, 8.0, 5.0, 5.0, 2.0, 8.0, 3.0, 2.0, 8.0, 6.0, 2.0, 9.0, 1.0, 4.0, 2.0, 7.0, 4.0, 1.0, 9.0, 6.0, 7.0, 4.0, 7.0, 4.0, 9.0, 5.0, 6.0, 4.0, 9.0, 1.0, 3.0, 5.0, 8.0, 6.0, 7.0, 1.0, 6.0, 7.0, 7.0, 3.0, 7.0, 6.0, 2.0, 1.0, 5.0, 4.0, 3.0, 6.0, 3.0, 3.0, 4.0, 5.0, 2.0, 8.0, 1.0, 8.0, 9.0]
Vecino prometedor con coste 128 encontrado: [3.0, 7.0, 8.0, 9.0, 1.0, 9.0, 5.0, 8.0, 5.0, 5.0, 2.0, 8.0, 3.0, 2.0, 8.0, 6.0, 2.0, 9.0, 1.0, 4.0, 2.0, 7.0, 4.0, 1.0, 9.0, 6.0, 7.0, 4.0, 7.0, 4.0, 9.0, 5.0, 6.0, 4.0, 9.0, 1.0, 3.0, 5.0, 8.0, 6.0, 7.0, 1.0, 6.0, 2.0, 7.0, 3.0, 7.0, 6.0, 2.0, 1.0, 5.0, 4.0, 3.0, 6.0, 3.0, 3.0, 4.0, 5.0, 2.0, 8.0, 1.0, 8.0, 9.0]
Vecino prometedor con coste 126 encontrado: [3.0, 7.0, 8.0, 9.0, 1.0, 9.0, 5.0, 8.0, 5.0, 5.0, 2.0, 1.0, 3.0, 2.0, 8.0, 6.0, 2.0, 9.0, 1.0, 4.0, 2.0, 7.0, 4.0, 8.0, 9.0, 6.0, 7.0, 4.0, 7.0, 4.0, 9

# Referencias empleadas

- Documentación de *numpy*: https://numpy.org/doc/
- Documentación de *pandas*: https://pandas.pydata.org/docs/
- Complejidad de la función *sum()* de Numpy: https://stackoverflow.com/questions/20135093/what-is-the-time-complexity-of-sum-in-python
- Complejidad de otras funciones de Numpy: https://stackoverflow.com/questions/52201990/is-there-a-list-of-big-o-complexities-for-the-numpy-library
- Complejidad de la función *shuffle* del módulo *random*: https://stackoverflow.com/questions/9371079/python-shuffle-algorithm-performance

# Indicaciones para trabajos futuros
- Para los algoritmos de Búsqueda Aleatoria y Búsqueda Local, las permutaciones generadas no contemplan los valores repetidos. Se están considerando $n!$ permutaciones en vez de $\frac{n!}{k!^M}$. Debería diseñarse una función de permutación que tenga en cuenta estas repeticiones para reducir el espacio de soluciones.
- Algunos algoritmos (especialmente el voraz) pueden fallar si el número de tomas totales no es divisible entre el número máximo de tomas por día (es decir, si $n/k$ no es exacta). Esto obliga a añadir un día extra donde no se llega al máximo de tomas posible, lo cual puede empujar al algoritmo a "inventar" índices adicionales que no encajan con el tamaño de los vectores ($n$).
- Debido a que el algoritmo voraz y sus derivados son totalmente deterministas, sería interesante añadir algo de diversificación (por ejemplo, implementando un algoritmo GRASP) para aprovechar su gran velocidad al tiempo que se permite más flexibilidad a la hora de explorar las soluciones.